File: client_macos_swift/Sources/VoiceTyperApp/Core/AppConfig.swift
import Foundation

struct AppConfig: Codable {
    var server: ServerConfig
    var hotkey: HotkeyConfig
    var hotwordFiles: [String]
    var ui: UIConfig

    init(
        server: ServerConfig = .init(),
        hotkey: HotkeyConfig = .init(),
        hotwordFiles: [String] = [AppConstants.defaultHotwordsFileName],
        ui: UIConfig = .init()
    ) {
        self.server = server
        self.hotkey = hotkey
        self.hotwordFiles = hotwordFiles
        self.ui = ui
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.server = try container.decodeIfPresent(ServerConfig.self, forKey: .server) ?? .init()
        self.hotkey = try container.decodeIfPresent(HotkeyConfig.self, forKey: .hotkey) ?? .init()
        self.hotwordFiles = try container.decodeIfPresent([String].self, forKey: .hotwordFiles) ?? [AppConstants.defaultHotwordsFileName]
        self.ui = try container.decodeIfPresent(UIConfig.self, forKey: .ui) ?? .init()
    }

    enum CodingKeys: String, CodingKey {
        case server
        case hotkey
        case hotwordFiles = "hotword_files"
        case ui
    }
}

struct ServerConfig: Codable {
    var scheme: String
    var host: String
    var port: Int
    var timeout: Double
    var apiKey: String
    var llmRecorrect: Bool
    var streaming: Bool

    init(
        scheme: String = "http",
        host: String = "127.0.0.1",
        port: Int = 6008,
        timeout: Double = 60.0,
        apiKey: String = "",
        llmRecorrect: Bool = true,
        streaming: Bool = true
    ) {
        self.scheme = scheme
        self.host = host
        self.port = port
        self.timeout = timeout
        self.apiKey = apiKey
        self.llmRecorrect = llmRecorrect
        self.streaming = streaming
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let rawScheme = try container.decodeIfPresent(String.self, forKey: .scheme) ?? "http"
        self.scheme = ServerConfig.normalizeScheme(rawScheme)
        self.host = try container.decodeIfPresent(String.self, forKey: .host) ?? "127.0.0.1"
        self.port = try container.decodeIfPresent(Int.self, forKey: .port) ?? 6008
        self.timeout = try container.decodeIfPresent(Double.self, forKey: .timeout) ?? 60.0
        self.apiKey = try container.decodeIfPresent(String.self, forKey: .apiKey) ?? ""
        self.llmRecorrect = try container.decodeIfPresent(Bool.self, forKey: .llmRecorrect) ?? true
        self.streaming = try container.decodeIfPresent(Bool.self, forKey: .streaming) ?? true
    }

    /// HTTP / HTTPS — 用于 /health、POST /recognize。
    var httpScheme: String {
        scheme.lowercased() == "https" ? "https" : "http"
    }

    /// WS / WSS — 与 httpScheme 对应。
    var wsScheme: String {
        scheme.lowercased() == "https" ? "wss" : "ws"
    }

    private static func normalizeScheme(_ raw: String) -> String {
        let lowered = raw.lowercased()
        return (lowered == "https" || lowered == "wss") ? "https" : "http"
    }

    enum CodingKeys: String, CodingKey {
        case scheme
        case host
        case port
        case timeout
        case apiKey = "api_key"
        case llmRecorrect = "llm_recorrect"
        case streaming
    }
}

struct HotkeyConfig: Codable {
    var modifiers: [String]
    var key: String

    init(modifiers: [String] = [], key: String = "fn") {
        self.modifiers = modifiers
        self.key = key
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.modifiers = try container.decodeIfPresent([String].self, forKey: .modifiers) ?? []
        self.key = try container.decodeIfPresent(String.self, forKey: .key) ?? "fn"
    }

    var displayString: String {
        if key.lowercased() == "fn" {
            return "Fn🌐"
        }
        let parts = modifiers + [key]
        return parts.map { $0.uppercased() }.joined(separator: "+")
    }
}

struct UIConfig: Codable {
    var opacity: Double
    var width: Double
    var height: Double

    init(opacity: Double = 0.85, width: Double = 240, height: Double = 70) {
        self.opacity = opacity
        self.width = width
        self.height = height
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.opacity = try container.decodeIfPresent(Double.self, forKey: .opacity) ?? 0.85
        self.width = try container.decodeIfPresent(Double.self, forKey: .width) ?? 240
        self.height = try container.decodeIfPresent(Double.self, forKey: .height) ?? 70
    }
}

File: client_windows_native/Core/AppConfig.cs
using System.Collections.Generic;
using YamlDotNet.Serialization;

namespace VoiceTyper.Core;

internal sealed class AppConfig
{
    [YamlMember(Alias = "server")]
    public ServerConfig Server { get; set; } = new();

    [YamlMember(Alias = "hotkey")]
    public HotkeyConfig Hotkey { get; set; } = new();

    [YamlMember(Alias = "hotword_files")]
    public List<string> HotwordFiles { get; set; } = new() { Support.AppConstants.DefaultHotwordsFileName };

    [YamlMember(Alias = "ui")]
    public UIConfig UI { get; set; } = new();

    public AppConfig Clone() => new()
    {
        Server = Server.Clone(),
        Hotkey = Hotkey.Clone(),
        HotwordFiles = new List<string>(HotwordFiles),
        UI = UI.Clone(),
    };
}

internal sealed class ServerConfig
{
    [YamlMember(Alias = "host")]
    public string Host { get; set; } = "127.0.0.1";

    [YamlMember(Alias = "port")]
    public int Port { get; set; } = 6008;

    [YamlMember(Alias = "timeout")]
    public double Timeout { get; set; } = 60.0;

    [YamlMember(Alias = "api_key")]
    public string ApiKey { get; set; } = "";

    [YamlMember(Alias = "llm_recorrect")]
    public bool LlmRecorrect { get; set; } = true;

    [YamlMember(Alias = "streaming")]
    public bool Streaming { get; set; } = true;

    public ServerConfig Clone() => new()
    {
        Host = Host,
        Port = Port,
        Timeout = Timeout,
        ApiKey = ApiKey,
        LlmRecorrect = LlmRecorrect,
        Streaming = Streaming,
    };
}

internal sealed class HotkeyConfig
{
    [YamlMember(Alias = "modifiers")]
    public List<string> Modifiers { get; set; } = new() { "ctrl" };

    [YamlMember(Alias = "key")]
    public string Key { get; set; } = "f2";

    public HotkeyConfig Clone() => new()
    {
        Modifiers = new List<string>(Modifiers),
        Key = Key,
    };

    public string DisplayString
    {
        get
        {
            var parts = new List<string>();
            foreach (var m in Modifiers)
            {
                parts.Add(NormalizeModifierDisplay(m));
            }
            parts.Add(Key.ToUpperInvariant());
            return string.Join("+", parts);
        }
    }

    private static string NormalizeModifierDisplay(string m) => m.ToLowerInvariant() switch
    {
        "ctrl" or "control" => "Ctrl",
        "alt" or "option" => "Alt",
        "shift" => "Shift",
        "win" or "win_l" or "win_r" or "super" or "command" or "cmd" => "Win",
        _ => m,
    };
}

internal sealed class UIConfig
{
    [YamlMember(Alias = "opacity")]
    public double Opacity { get; set; } = 0.85;

    [YamlMember(Alias = "width")]
    public double Width { get; set; } = 320;

    [YamlMember(Alias = "height")]
    public double Height { get; set; } = 90;

    public UIConfig Clone() => new()
    {
        Opacity = Opacity,
        Width = Width,
        Height = Height,
    };
}
